[ENH] Datalad documentation, archiving The Way, updating information on Jekyll versions for GH page building (#80)
PennLINC/PennLINC.github.io: docs/DataWorkflows/DataCycle.md

---
layout: default
title: Data Cycle
parent: Data Workflows
nav_order: 1
has_toc: true
---
# The Data Cycle

The data cycle includes the following steps: 
1. Fetching raw data

> 3. Curation of datasets

Reviewer: ```suggestion
3. Dataset curation
```

2. Fetching the data onto the right cluster
3. Unzipping the data (as needed)
4. Processing the data!
5. Sharing the data

Each of these steps is detailed here: 

> # Fetching Your Data

Reviewer: I propose including datalad here. That's how I grab OpenNeuro datasets.

Author: added

```
## my username is <username>
scp -r path/to/your/data <username>@transfer:/path/on/pmacs
```

An alternative to `scp` is `rsync`, but that tends to have [more happening under the hood](https://stackoverflow.com/questions/20244585/how-does-scp-differ-from-rsync).

## Flywheel

On Flywheel, your data may already be in BIDS. In this case we recommend using Flywheel's export function `fw export bids`, or the export function provided by [`fw-heudiconv`](https://fw-heudiconv.readthedocs.io/en/latest/). We built the export function into `fw-heudiconv` because we wanted to have more flexibility in what BIDS data we could grab, including data that's not yet supported in the official BIDS spec. Admittedly though, downloading all of `fw-heudiconv` a lot of overhead for just the export function.

```
# with fw export bids
fw export bids <DESTINATION_DIRECTORY> --project <PROJECT_NAME> --subject <SUBJECT_FILTER>

# with fw-heudiconv
fw-heudiconv-export --project <PROJECT_NAME> --subject <SUBJECTS_FILTER> --session <SESSION_FILTER> --folders <LIST_OF_BIDS_FOLDERS>
```

Try `fw-heudiconv-export -h` for more info.

## Globus

> ##### Note: this section has not been tested in a long time

Reviewer: ```suggestion
{: .warning-title }
> Warning
>
> This section has not been tested in a long time.
```

Red admonition box to make it pop.

On PMACs:

0. Make sure you have access to the PULSE Secure VPN -- [remote.pmacs.upenn.edu](remote.pmacs.upenn.edu)

1. Log in to PMACs' dedicated node for Globus functionality:

```
# first ssh into bblsub for network access
ssh -y <username>@bblsub

# then from there, log onto the globus node
ssh -y <username>@sciglobus
```

2. Globus Connect Personal should be available. As above, use it to initialize an endpoint on a directory of your choice on PMACs. Specifically, you should run it as below so that it opens a GUI for logging in with an auto-generated token:

```
# this command will return a URL you can open in any browser and a token you can use to sign in
globusconnect -start &
```

3. Using a new or existing conda environment (see [here](https://pennlinc.github.io/docs/pmacs#logging-in-to-pmacs-lpc) for how to activate conda on PMACs), install the [CLI](https://docs.globus.org/cli/) using `pip` and login with `globus login`.

4. Visit [https://docs.globus.org/how-to/get-started/](https://docs.globus.org/how-to/get-started/) to access the File Manager, as in the Local Disk instructions, to start transferring data.

# Curating BIDS Datasets

> BIDS curation can be a frustrating process. This guide describes best practices for
> the curation process using a local filesystem. We divide this process into multiple
> stages.

Reviewer: ```suggestion
BIDS curation can be a frustrating process. 
This guide describes best practices for the curation process using a local filesystem. 
We divide this process into multiple stages.
```

I find version control on markdown files easier if I split lines based on sentences and commas. No need to adopt this if you don't like it.

```

[Download and install CuBIDS](https://bids-bond.readthedocs.io/en/latest/index.html). Note that this environment must be activated for the rest of the steps.

You'll also need to install BIDS Validator:

```shell
$ conda install nodejs
$ npm install -g bids-validator
$ which bids-validator #to make sure it installed
```

## Stage 0: Organization

The process begins when data arrives on one of our servers. While it may not
be possible to document everything that has happened to the data prior to its
arrival to our server, the goal of the organization step is to document as
much as we can about the process of receiving the data, and any initial
changes we might make to it. The person in charge of organizing a data set
should document any changes they make in a *Data Narrative* along with any
relevant scripts used to a git-tracked repository.

Create a project root directory. Initially we suggest keeping one copy of your
initial data that will remain untouched as a backup during early stages of
the workflow. This is not necessary if your data is archived elsewhere online.
In the project root create a directory structure similar to this:

> ```

Reviewer: I think it would be good to link out to [`paper-template`](https://github.com/PennLINC/paper-template) for this. If that template doesn't have a good proposed organization, we can change it.

# the actual directory structure can
# vary depending on the needs of your project

project
├── original_data
│   ├── sub-1
│   ├── sub-2
│   ├── ...
│   └── sub-N
└── curation
    ├── code
    |   ├── sandbox                # a place for untracked files
    │   └── ProjectGithub          # the home for all your scripts, tracked by github
    |       └── DataNarrative.md   # the data narrative file
    └── BIDS
```

The `original_data` directory should be an unchanged copy of the original data. You can
get a template Data Narrative file with:

```shell
$ wget https://raw.githubusercontent.com/PennLINC/RBC/master/Data_Narrative_Template.md
```

Next, add tracking to your project for scripts to be shared and tracked. 
data:

```
project
├── original_data
│   ├── sub-1
│   ├── sub-2
│   ├── ...
│   └── sub-N
└── curation
    ├── code
    |   ├── sandbox
    │   └── ProjectGithub         # add git here
    |       └── DataNarrative.md
    └── BIDS
```

You can add git like so:

```bash
$ git init
```

> And remember, if you explicitly want to ignore something from tracking (with either `git` or `datalad`), add that path to a `.gitnignore` file ([see here](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) and [here](https://handbook.datalad.org/en/latest/beyond_basics/101-179-gitignore.html)).

Reviewer: ```suggestion
And remember, if you explicitly want to ignore something from tracking (with either `git` or `datalad`), add that path to a `.gitignore` file ([see here](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) and [here](https://handbook.datalad.org/en/latest/beyond_basics/101-179-gitignore.html)).
```

> ### Add NIfTI information to the sidecars

Reviewer: Probably best to link out to CuBIDS's documentation for this, since CuBIDS _should_ have this info in one of its tutorials (if it doesn't, we should add it).

Author: Yep, we will have to add it

# Unzipping your data

Occasionally, data may be zipped. More information about this can be found [here](/PennLINC.github.io/docs/DataWorkflows/FetchingYourData.md)!

# Processing your data
[BABS](https://pennlinc-babs.readthedocs.io/en/latest/) is a quick and easy tool for processing pipelines via [Datalad](./Datalad.md). The documentation is thorough, with information on setting up, installing, and running BABS on data. An example walkthrough is also available via the documentation. 

# Sharing your data
{: .no_toc}

Collaboration is an important part of our work. Sharing the input data and outputs/derivatives of a project is a critical step in the scientific process, and so doing it both accurately and efficiently is a high priority. Overall, remember that this step can be a strain, especially when dealing with external collaborators or systems that are foreign to you. Be patient, and remember that we have to balance accuracy/reproducibility with efficiency/speed.

Below are a few example scenarios and what best recommendations we have for sharing data, particularly outputs from BABS

> ## Datalad to Datalad

Reviewer: Can you add a section on Datalad to OpenNeuro? See #60. Can be left empty here, but will remind us to fill it out at some point.

Author: added the section + linked to docs on how to do it
